MAINT: Refactor UFunc core to use NEP 43 style dispatching/promotion

seberg opened the pull request:
The first (big) chunk of a multi-step process to update all ufuncs
to NEP 43 style ufuncs.

This limits itself to normal ufunc calls and sets ufunc.at and
all reduce-like operations to the side.  These would add another
level of complexity.

---

This is the split-off removing some reductions (and some complexity that only reductions need to pass the test-suite).  It is still meant to be rebased after gh-19371 and gh-19381 are merged.

This is still big, and the `ufunc_object.c` changes are complex (the others are complicated due to their size and being fully new infrastructure).  So I have to think a bit more if I can split it up...

There are still multiple bigger areas that have to be attacked separately, but are hard to unravel completely (although maybe possible in a similar way to the hash-map):

1. `array_method.c` changes are refactors to the masked strided-loop to match the new ufunc signature.
2. `numpy/core/src/umath/legacy_array_method.c` is the logic that wraps the current ufunc inner-loops into the new `ArrayMethod` concept.
   * They even still call the `ufunc->legacy_inner_loop_selector` for absolute ABI compatibility (hopefully we can drop that quickly).
3. `dispatching.c` is the trickiest probably, since implements the multiple dispatching logic, and is the easiest to include some subtle errors.
4. The changes to `ufunc_object.c`, which largely split into the two "big" categories:
    * dispatching (e.g. with `signature` refactoring) and special handling to support value-based casting.
    * using the ArrayMethod's strided-loops by passing through `ufuncimpl` everywhere.

seberg commented:
@mattip and @mhvk do you think that this may be reviewable already?  The main thing to split off would be the promotion code, I think.  But I am not quite sure how to best test it on its own.

It is definitely still too big to really think through subtleties, although I am not sure that is quite reachable (but it is definitely better that the other small chunks are not here anymore).

seberg commented:
So the one thing I can think of is to implement `ufunc._get_implementation(...) -> ArrayMethod` to write some "unit tests".  That will remove most changes to the `ufunc_object.c`, but it will also add some additional complexity...

So, I am unsure it would actually help at all with review.  We could take some time to walk through the PR (or maybe better chunks of it) if that helps?

One tricky part is the multiple-dispatching:  It is not really used yet, since we will always only find exact matches.  If we do not, we have to use promotion; and that is still the legacy version.  So it is hard to be sure that the logic is fully right, bugs might only be flushed out when we start using promoters (e.g. new ufuncs not using the legacy loops at all, or add them to existing ones to avoid the legacy fallback).

seberg commented:
I am planning to meet on this with Matti tomorrow at around 3pm UTC to kick-off the review process.  In case anyone is interested.

mhvk commented:
@seberg - this PR is definitely on my list to review, but I've have had a bit of trouble finding the required larger amount of time. Will definitely be good if you can discuss with @mattip and perhaps others (I cannot make it, sadly).

mattip commented on numpy/core/include/numpy/ufuncobject.h:
```diff
@@ -211,6 +211,13 @@ typedef struct _tagPyUFuncObject {
         /* Identity for reduction, when identity == PyUFunc_IdentityValue */
         PyObject *identity_value;
 
+        /* New private fields related to dispatching */
```
```suggestion
        /* New in NPY_API_VERSION 0x0000000F and above */
        /* New private fields related to dispatching */
```

mattip commented on numpy/core/include/numpy/ufuncobject.h:
```diff
@@ -211,6 +211,13 @@ typedef struct _tagPyUFuncObject {
 
+        /* New private fields related to dispatching */
+        void *_dispatch_cache;
+        /*
+         * Currently, just a list, but that can never allow deletion (unless
+         * elements that may get deleted again are deleted occasionally).
+         */
```
```suggestion
        /*
         * Currently, just a list, but that can never allow deletion (unless
         * elements that may get deleted again are deleted occasionally).
         */
```
```suggestion
        /* A PyListObject of (dtype, array-loop-function)  */
```

seberg replied:
Adopted as `/* A PyListObject of `(tuple of DTypes, ArrayMethod/Promoter)` */`

The TODO is a bit of a long story and very tricky.  So tending to skip it...

(The problem is what happens with registered or cached ufunc loops that become inaccessible if you *delete* a `DType` class.  But, I have yet to see a multiple-dispatching system dealing with that.  And I am not even sure that Python **can** deal with it; I think the GC is missing a feature necessary to do it properly anyway.)

mattip commented on numpy/core/src/multiarray/array_method.c:
```diff
@@ -819,39 +817,36 @@ unmasked_ufunc_loop_as_masked(
-        strided_loop(dataptrs, &subloopsize, strides, innerloopdata);
+        if (strided_loop(context,
+                dataptrs, &subloopsize, strides, strided_loop_auxdata) != 0) {
+            return -1;
```
Maybe store the return value and compare it to 0, then return the value on error. This might make it future-proof

mattip commented on numpy/core/src/multiarray/array_method.c:
```diff
@@ -865,18 +860,13 @@ PyUFunc_DefaultMaskedInnerLoopSelector(PyUFuncObject *ufunc,
+    if (context->method->get_strided_loop(context,
+            aligned, 0, fixed_strides,
+            &data->unmasked_stridedloop, &data->unmasked_auxdata, flags) < 0) {
+        PyMem_Free(data);
```
return value got lost

seberg replied:
This one is fine I think, I don't see `get_strided_loop` getting a meaningful return value anytime soon?

mattip commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -0,0 +1,697 @@
+ * - ufunc._loops: A list of all ArrayMethods and promoters, it contains
+ *   tuples `(dtypes, ArrayMethod)` or `(dtypes, promoter)`.
+ * - ufunc._dispatch_cache: A cache to store previous promotion and/or
+ *   dispatching results.
```
Should this have an upper bound?

seberg replied:
> Should this have an upper bound?

Yes, although it is probably hard to run into an issue.  I will mark this as resolved though, moving to the issue gh-19544, not sure how crucial it is, but attacking it in this PR is not going to help I think.

mattip commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -0,0 +1,697 @@
+    PyObject *DType_tuple = PyTuple_GetItem(info, 0);
+
+    PyObject *loops = ufunc->_loops;
+    Py_ssize_t length = PyList_Size(loops);
```
Check that length matches nin

mhvk commented on numpy/core/src/umath/legacy_array_method.c:
```diff
@@ -0,0 +1,259 @@
+
+
+/*
+ * This just seems to tricky to make work correctly, there are two main
```
to -> too?

mhvk reviewed (commented):
Very small comments before trying to dive in...

mhvk commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -0,0 +1,697 @@
+ * steps:
+ *
+ * 1. Override any `operand_DTypes` from `signature`.
+ * 2. Check if the new `operand_Dtypes` is cached (got to 4. if it is)
```
"if it is, go to 4"

(And *really* nice to have this explanation on top!!)

mhvk commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -0,0 +1,697 @@
+        PyArray_DTypeMeta *op_dtypes[], PyObject **out_info)
+{
+    int nin = ufunc->nin, nargs = ufunc->nargs;
+    /* Use new style type resolution has to happen... */
```
Not quite sure what is intended with this comment.

mhvk commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -0,0 +1,697 @@
+                 * (e.g. add/multiply ensure high precision).
+                 * Continuing here matches promoters: those can deal with that.
+                 * If we allow this path for ArrayMethod, the person
+                 * registering will have it works sense for the ufunc, a
```
??  Did you mean something like "must have made sure it makes sense"? I don't quite understand the example.

seberg replied:
Changing to:
```
                 * If the input was not specified, this is a reduce-like
                 * operation.  Some ufuncs may have non-trivial promotion
                 * (e.g. add/multiply ensure high precision).
                 * Continuing here allows a promoter to handle reduce-like
                 * operations explicitly.  E.g. upcasting boolean to integer.
                 * TODO: The `!resolver_dtype->abstract` currently ensures that
                 *       this is a promoter.  If we allow ArrayMethods to use
                 *       abstract DTypes, we may have to reject it here or the
                 *       ArrayMethod has to implement the reduce promotion.
```
there are probably still some more loose ends with the possibility of ArrayMethod's being registered for abstract DTypes (but generally, I think it may be useful).

seberg replied:
Nvm, seeing the next comment, it now is:
```
                 * If an input was not specified, this is a reduce-like
                 * operation: reductions use `(operand_DType, NULL, out_DType)`
                 * as they only have a single operand.  This allows special
                 * reduce promotion rules useful for example for sum/product.
                 * E.g. `np.add.reduce([True, True])` promotes to integer.
                 *
                 * Continuing here allows a promoter to handle reduce-like
                 * promotions explicitly if necessary.
                 * TODO: The `!resolver_dtype->abstract` currently ensures that
                 *       this is a promoter.  If we allow ArrayMethods to use
                 *       abstract DTypes, we may have to reject it here or the
                 *       ArrayMethod has to implement the reduce promotion.
```
puh :)

mhvk commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -0,0 +1,697 @@
+            }
+            if (given_dtype == NULL) {
+                /*
+                 * If the (input) was not specified, this is a reduce-like
```
Why would the input not be specified for a reduction?

seberg replied:
I think I answered it in the comment, but to be sure.  The idea is to use the dtypes: `(NULL, arr.dtype, NULL)` (or `(arr.dtype, NULL, NULL)`) for reductions, that way the promotion usually knows that a reduction is happening, and can special case it.

That helps mainly so that the logic of returning `np.int_` for `np.array([True, True]).sum()` can be defined as a "promotion" (and I think I needed it also to special case the logical "reductions").  That said, it may well be that we will have to make this more explicit.

mhvk commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -0,0 +1,697 @@
+                 */
+                continue;
+            }
+            int subclass = PyObject_IsSubclass(
```
It seems strange that this line would not be covered. Is it true? Is it because we are only testing with numpy types right now?

seberg replied:
The whole subclass logic makes only sense if we have abstract DTypes.  Those do not really exist yet.  (Some do exist, but they only ever take part in value-based promotion, and currently always uses the legacy fallback always.)

It will be used if we have both:
* Promoters
* DTypes beyond "legacy" ones.

mhvk commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -0,0 +1,697 @@
+             * In all cases, we give up resolution, since it would be
+             * necessary to compare to two "best" cases.
+             */
+            int unambiguous_equivally_good = 1;
```
should this be `unambigous_equally_good`?

mhvk commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -0,0 +1,697 @@
+            for (Py_ssize_t i = 0; i < nargs; i++) {
+                int best;
+
+                /* Whether this (normally output) dtype was specified at all */
```
I'm confused how this can happen at all, given that you assert given_dtype != PyNone above (l.158)

seberg replied:
Hmmm, a bit confusing.  In the ufunc code, I currently use `PyArray_DTypeMeta *DTypes[NPY_MAXARGS]` where `NULL` represents missing.  But this comes from a tuple, where `None` represents missing.

So one would be `NULL`, while the other would be `None` to represent the same thing.

mhvk commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -0,0 +1,697 @@
+                    /* equivalent, so this entry does not matter */
+                    continue;
+                }
+                if (is_not_specified) {
```
Yet this code seems covered, so I must be missing something... Or is it not actually covered?

seberg replied:
Yeah, it is covered. The new TODO, is a bit clearer maybe.

The uncovered parts have to do with abstract DTypes mainly (which is different from unspecified).  That means all the `abstract` and `subclass` paths are uncovered.

mhvk commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -0,0 +1,697 @@
+            if (current_best == -1) {
+                /*
+                 * TODO: It would be nice to have a "diagnostic mode" that
+                 *       informs if this happens! (An immediate error current
```
current -> currently

mhvk commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -0,0 +1,697 @@
+
+
+/*
+ * Used for the legacy fallback promotion when `signature` or `dtype` is
```
Maybe tell what the function does!

mhvk commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -0,0 +1,697 @@
+        PyArray_DTypeMeta *operation_DTypes[], int *out_cacheable)
+{
+    int nargs = ufunc->nargs;
+    PyArray_Descr *out_descrs[NPY_MAXARGS];
```
Just use `= {NULL}` instead of `memset`? Or does that not guarantee setting to 0 at each entry?

seberg replied:
Probably happened in some micro-optimization frenzy when comparing old and new speeds.  NULL'ing out `32*8 == 256` bytes is a bit more expensive than the `3*8 == 24` that we usually need (I guess since it touches more than one cache line).

Changing it for now.

mhvk commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -0,0 +1,697 @@
+                PyTuple_GET_ITEM(info, 1), &PyArrayMethod_Type)) {
+            /*
+             * Cache the new one.  NOTE: If we allow a promoter to return
+             * a new ArrayMethod, we should also cache such a promoter also.
```
"we should also cache that information."

seberg replied:
Removed the note for now (modified the comment).  Not really the right place for it, and right now there is no good place, since promoters are not really a thing in this version.

mhvk commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -0,0 +1,697 @@
+
+
+/*
+ * The central entry-point for the promotion and dispatching machinery.
```
This is the second routine which is called the central entry point!

mhvk commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -0,0 +1,697 @@
+
+    PyArrayMethodObject *method = (PyArrayMethodObject *)PyTuple_GET_ITEM(info, 1);
+
+    /* Fill in the signature with the signature that we will be working with */
```
"with the dtypes that we will be working with"?

mhvk reviewed (commented):
Some first review of `dispatching.c`. One difficulty is that it is not clear what is and is not active code right now...

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -924,30 +935,67 @@ convert_ufunc_arguments(PyUFuncObject *ufunc,
     PyObject *obj;
 
     /* Convert and fill in input arguments */
+    int all_scalar = 1;
```
I've wondered more generally, but when I was contributing, I was at some point told to use `npy_bool` for cases like this. I think this gives a bit of clarity, though really it doesn't matter much!

seberg replied:
Hmm, yeah... We even can use C99 bools.  Mainly a missing habit for me, will change it here (but not everywhere yet).

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -924,30 +935,67 @@ convert_ufunc_arguments(PyUFuncObject *ufunc,
+            }
+        }
+        out_borrowed_op_DTypes[i] = NPY_DTYPE(PyArray_DESCR(out_op[i]));
+        if (!out_borrowed_op_DTypes[i]->legacy) {
```
I guess the fact that this is not covered answers my question elsewhere, that there are no tests yet for the new path. I wonder if there shouldn't be at least some simple ones!?

seberg replied:
I am operating under the thought of getting NumPy ready first and writing tests when this becomes closer to public API, or as we might use it internally even.
That could be as part of an "example"  DType.  For me that just seemed easier to keep things moving, even if it is bound to lead to some fixups being necessary later.

But if it seems better or helps with understanding the code, I can try to start adding them here I guess.

That said, I do not think I can cover *all* branches yet and leaving them out is worse IMO.  But, I could cover some of using the private API.  (For one, I omitted promotion, which would be good to have for tests.)

mhvk replied:
I think it would be very useful to have tests for things that already work, if only to help review! For my own code, I find that the tests sometimes become better documentation than the docstrings... Plus, they invariably smoke out errors. And, as you note, they end up being useful as examples in documentation. So, yes, if you can, please do add some tests!

seberg replied:
Well, working on it (just pushed, will probably fail the tests).  I need casting etc. as well (which somewhat would work without ufuncs)...  Unfortunately it is 660 lines of C code, although mostly straight forward probably.

The idea was to use a "scaled" float/double (not really useful, but a bit like a unit).

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -924,30 +935,67 @@ convert_ufunc_arguments(PyUFuncObject *ufunc,
-            out_op[i] = (PyArrayObject *)PyArray_FromArray(obj_a, NULL, 0);
+            out_op[i] = (PyArrayObject *)obj;
+            Py_INCREF(out_op[i]);
+        }
```
I find the twisting in and out a bit confusing; it might be clearer to just do the scalar check here and then slightly duplicate it below for the non-array case. But not a big deal.

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -924,30 +935,67 @@ convert_ufunc_arguments(PyUFuncObject *ufunc,
+            any_scalar = 1;
+        }
+        if ((PyObject *)out_op[i] == obj) {
+            continue;  /* input was an array: no need to check for pyscalar */
```
See comment above. If not refactoring, I'd write this as != and put all the next type-checking statements inside.

seberg replied:
Adopted for now, but you are right, there may well be a better way to write this. Coming "back" to the original object is a bit strange.

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2034,13 +2090,11 @@ PyUFunc_GeneralizedFunctionInternal(PyUFuncObject *ufunc,
     npy_intp inner_dimensions[NPY_MAXDIMS+1];
     /* The strides which get passed to the inner loop */
     npy_intp *inner_strides = NULL;
+    NpyAuxData *auxdata = NULL;
```
Maybe add comment for this line too (though it is pretty obvious, I guess).

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2057,6 +2111,11 @@ PyUFunc_GeneralizedFunctionInternal(PyUFuncObject *ufunc,
 
     NPY_UF_DBG_PRINT1("\nEvaluating ufunc %s\n", ufunc_name);
 
+    if (validate_casting(ufuncimpl,
```
I've slightly lost track, but it seems `validate_casting` is called in quite a few places. Are there no duplications? Could one assume it is OK here?

Though if I trust the coverage, then obviously this branch is triggered at least sometime!

seberg replied:
It is called once for normal ufuncs, and once for gufuncs.  The "odd" thing is the trivial-loop special case/preparation.  That does not use it, but falls back to the slow path when in doubt.

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -5105,6 +5223,28 @@ PyUFunc_FromFuncAndDataAndSignatureAndIdentity(PyUFuncGenericFunction *func, voi
+    char *curr_types = ufunc->types;
+    for (int i = 0; i < ntypes * (nin + nout); i += nin + nout) {
+        /*
+         * Add all legacy wrapping loops here. This is normally necessary, but
```
normally *not* necessary?

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -5105,6 +5223,28 @@ PyUFunc_FromFuncAndDataAndSignatureAndIdentity(PyUFuncGenericFunction *func, voi
+        PyArray_DTypeMeta *op_dtypes[NPY_MAXARGS];
+        for (int arg = 0; arg < nin + nout; arg++) {
+            op_dtypes[arg] = PyArray_DTypeFromTypeNum(curr_types[arg]);
+            Py_DECREF(op_dtypes[arg]);  /* That DType can't be deleted... */
```
Took a bit of time to grok the comment (here and below): I think the point is that the DType cannot be deleted in between here and where the loop is added, so we might as well decref it now rather than have a separate decref later, correct?

seberg replied:
Tweaked the comment to `These DTypes are immortal and adding INCREFs: so borrow it` (if still not obvious, will just explicitly decref them later).

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -5347,11 +5489,59 @@ PyUFunc_RegisterLoopForType(PyUFuncObject *ufunc,
+        goto fail;
+    }
+    /*
+     * We add the loop to the new list of all loops and promoters.  If the
```
Why "new list"? Just "list"?

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -5347,11 +5489,59 @@ PyUFunc_RegisterLoopForType(PyUFuncObject *ufunc,
+    }
+    /*
+     * We add the loop to the new list of all loops and promoters.  If the
+     * equivalent loop was already added, skip this.  If a different loop is
```
I'm a bit confused why we skip if the equivalent loop was already added. Shouldn't we replace? Or check that the implementation is identical? Or is that to be done later so that it replaces the still existing registration?

seberg replied:
Maybe this comment now makes it clear?:

>We add the loop to the list of all loops and promoters.  If the equivalent loop was already added, skip this.
> Note that even then the ufunc is still modified: The legacy ArrayMethod already looks up the inner-loop from the ufunc (and this is replaced below!).
> If the existing one is not a legacy ArrayMethod, we raise currently: A new-style loop should not be replaced by an old-style one.

although, we may have to allow that replacement with a deprecation-warning before too long, so that we don't break those replacing our loops but can use them internally anyway.

mhvk reviewed (commented):
Now some comments on the changes in `ufunc_object.c`. As those will make clear, I somewhat lost the thread and wasn't quite sure what is to come in the future still. Hopefully, still some use...

mhvk commented:
p.s. Now looked at everything and am slightly confused in some places. But just wanted to make clear that in most places the code is a *lot* clearer than it was before!

charris commented:
close/reopen

mattip commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -924,21 +935,60 @@ convert_ufunc_arguments(PyUFuncObject *ufunc,
-                                    NULL, 0, 0, 0, NULL);
+            /* Convert the input to an array and check for special cases */
+            out_op[i] = (PyArrayObject *)PyArray_FromAny(obj, NULL, 0, 0, 0, NULL);
+            if (out_op[i] == NULL) {
```
This is a bit awkward. The argument is converted to an array, but if it is one of the special objects (python long, float, complex) the `out_op_DTypes` is marked with the type, and later on, value-based promotion is used. Wouldn't it be better to avoid the conversion here, since the value-based promotion can use the original object anyway?

seberg replied:
After trying to figure out a way to make this nicely, I opted to just drop it for now with a huge comment.  It probably makes more sense to defer this till after we attempt getting rid of value-based casting for good (which I assume I can try very soon if this moves well!)

That said, it is probably important to get this right (or warn about it) when we make DTypes public, so added it to the issue.

mattip commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -1243,46 +1302,88 @@ try_trivial_single_output_loop(PyUFuncObject *ufunc,
     NPY_END_THREADS;
+    NPY_AUXDATA_FREE(auxdata);
+
+    if (flags & NPY_METH_NO_FLOATINGPOINT_ERRORS || res < 0) {
+        return res;
+    }
+    /* NOTE: We could check float errors even when `res < 0` */
+    const char *name = ufunc_get_name_cstr((PyUFuncObject *)context->caller);
+    return _check_ufunc_fperr(errormask, extobj, name);
```
```suggestion
    if (res < 0) {
        return res;
    }
    if (flags & NPY_METH_NO_FLOATINGPOINT_ERRORS) {
        /* NOTE: We could check float errors even when `res < 0` */
        const char *name = ufunc_get_name_cstr((PyUFuncObject *)context->caller);
        return _check_ufunc_fperr(errormask, extobj, name);
    }
    return 0;
```

seberg replied:
Actually, decided to go with the:
```
    if (res == 0 && !(flags & NPY_METH_NO_FLOATINGPOINT_ERRORS)) {
        /* NOTE: We could check float errors even when `res < 0` */
        const char *name = ufunc_get_name_cstr((PyUFuncObject *)context->caller);
        res = _check_ufunc_fperr(errormask, extobj, name);
    }
    return res;
```
pattern.  That is a bit flexible to use unconditionally.

mattip commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2427,29 +2499,28 @@ PyUFunc_GeneralizedFunctionInternal(PyUFuncObject *ufunc,
-        goto fail;
+    if (retval == 0 && !(flags & NPY_METH_NO_FLOATINGPOINT_ERRORS)) {
+        /* NOTE: We could check float errors even when `res < 0` */
+        retval = _check_ufunc_fperr(errormask, extobj, ufunc_name);
```
Should be consistent with the code pattern for this same construct in the other loops

mattip commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -4446,15 +4469,76 @@ _get_normalized_typetup(PyUFuncObject *ufunc,
+        retval = 0;
+    }
+    else {
+        /* Rare fall-back to the legacy resolver to pass `operands` */
```
Change comment, instead of "Rare" put under which conditions this will be used: it is for datetime?

seberg commented on doc/source/reference/c-api/types-and-structures.rst:
```diff
@@ -957,6 +958,12 @@ PyUFunc_Type and PyUFuncObject
+            called and not be used in the future.
+            See ref:`NEP 41 <NEP41>` and ref:`NEP 43 <NEP43>`
+
        A function which returns an inner loop. The ``legacy`` in the name arises
```
Made the `.. note` into `.. deprecated` and just deleted it.  Its not used, so users don't really need to know what it does (it is not like anyone uses it anyway).

seberg commented:
I think I addressed all comments now.  After talking a bit today with Matti, I have removed:

1. The multiple-dispatching paths related to abstract DTypes, so there should be much better coverage there now (and big holes for the future to fill again :))
2. The "weak" DType special case, which we will probably need to allow for new-style promotions and need work-around for. (put into the tracking issue)

Both probably must happen pretty soon, but for the sake of review here it is probably easier to just skip it for now...

I *removed* the "test" again, though.  It is just much too long for almost no additional coverage at all. The one actual fix due to it is that I forgot to never use the legacy promotion when a new-style DType is involved (not that it doesn't work, but it seems undesirable).

The test (with some addition) may be helpful however be helpful to add promotion *before* the second part adding reductions here!

I will definitely open a PR for the test, the only question is there is much reason to put it in first. I don't really think so, now that the most "tricky" uncovered parts should be gone?


Luckily codecov fixed their bug, so the output is a bit more useful again.  I still expect some uncovered paths in the original matching code, but otherwise it is hopefully good now.

seberg commented:
@mhvk would you have time to have another look at this?  If it helps, happy to do that together as well whichever time suits you.  I pulled out most of the "controversial" things.

My current plan would be to put this in largely as is, do the "test" (a scaled floating point DType) next.  Then add simple promotion tests based on it.
After that, reductions will hopefully be a bit more straight forward, since they require some promotion addition.

mattip commented:
Did you run a test for memory leaks?

mattip commented on numpy/core/src/multiarray/array_method.c:
```diff
@@ -819,39 +817,37 @@ unmasked_ufunc_loop_as_masked(
- *
- * Returns 0 on success, -1 on error.
+ * Identical to the `get_loop` functions and wraps it.  This adds support
+ * to a boolean mask being passed in as a last, additional, operand.
```
Not clear: where is the boolean mask passed in. Do we even support boolean masks? Maybe leave this wrapper for a future PR when we add that support

mhvk commented:
@seberg - I'm observing this week, but see whether there is a bit of time. Otherwise, don't wait for my review; overall, it already looked good and it is very good to do it a bit more in pieces.

seberg commented:
> Did you run a test for memory leaks?

I had run one earlier, I ran the `ufunc` and `umath` tests through valgrind now (found a small error during ufunc registration).  I will do a more complete run eventually, but I think this should be sufficient to catch any serious leaks.

> Not clear: where is the boolean mask passed in.

Modified that docstring to mention the ufunc `where=...` keyword argument, that hopefully makes this clear?
